controllers/metal3.io/host_state_machine.go

```diff
@@ -79,13 +79,14 @@ func recordStateEnd(info *reconcileInfo, host *metal3v1alpha1.BareMetalHost, sta
 	return
 }
 
-func (hsm *hostStateMachine) ensureProvisioningCapacity(info *reconcileInfo) actionResult {
-	hasCapacity, err := hsm.Provisioner.HasProvisioningCapacity()
+func (hsm *hostStateMachine) ensureCapacity(info *reconcileInfo) actionResult {
+	hasCapacity, err := hsm.Provisioner.HasCapacity()
 	if err != nil {
-		return actionError{errors.Wrap(err, "failed to get hosts currently being provisioned")}
+		return actionError{errors.Wrap(err, "failed to get hosts currently being (de)provisioned")}
 	}
+
 	if !hasCapacity {
-		return recordActionDelayed(info)
+		return recordActionDelayed(info, hsm.NextState)
```

Comment:
This doesn't look fully correct. Only in one case the check is performed against `hsm.NextState`, while in two cases the `info.host.Status.Provisioning.State` it's used (and in one none at all).

Reply:
@andfasano I am not sure how you want to handle this.

Generally passing `hsm.NextState` to `recordActionDelayed()` is ok with you?

Also I did a bit for internal fixing (addressing also previous comment) in `recordActionDelayed()` so your point could be solved by rebased PR

Reply:
Sure, let my clarify my point with some additional details. I think that in `checkDelayedHost` the current host provisioning state should be passed to `recordActionDelayed`, while in `updateHostStateFrom` it's ok to use `hsm.NextState`. Even though technically speaking `hsm.NextState` is equivalent to the `currentState` when evaluated in `checkDelayedHost`, I think that for the sake of clarity - and moreover to avoid future issues in case of refactoring - it's better to make it clear and explicit (so that the current temporal dependency will be removed)

Reply:
So removing `hsm.NextState` and only using `info` to detect state?

> Even though technically speaking hsm.NextState is equivalent to the currentState when evaluated in checkDelayedHost

Hmm, this was not obvious to me, so I used `hsm.NextState`, removing then makes sense, also I would like to add a comment about it in code to make it more clearer for other contributors.

Reply:
> So removing `hsm.NextState` and only using `info` to detect state?

Unfortunately I don't think it could be sufficient, you'll still need an extra parameter for the state in the `ensureCapacity` I guess